src/music/midifile.cpp

```diff
@@ -440,6 +452,338 @@ bool MidiFile::LoadFile(const char *filename)
+	{
+		*block.data.Append() = b1;
+		*block.data.Append() = b2;
+		*block.data.Append() = b3;
+	}
+
+	/**
+	 * Construct a TTD DOS music format decoder.
+	 * @param songdata Buffer of song data from CAT file, ownership remains with caller
+	 * @param songdatalen Length of the data buffer in bytes
+	 * @param target MidiFile object to add decoded data to
+	 */
+	MpsMachine(const byte *data, size_t length, MidiFile &target)
+		: songdata(data), songdatalen(length), target(target)
+	{
+		uint32 pos = 0;
+		int loopmax;
+		int loopidx;
+
+		/* First byte is the initial "tempo" */
+		this->song_tempo = this->songdata[pos++];
+
+		/* Next byte is a count of callable segments */
+		loopmax = this->songdata[pos++];
+		for (loopidx = 0; loopidx < loopmax; loopidx++) {
+			/* Segments form a linked list in the stream,
+			 * first two bytes in each is an offset to the next.
+			 * Two bytes between offset to next and start of data
+			 * are unaccounted for. */
+			this->segments.push_back(pos + 4);
+			pos += FROM_LE16(*(const int16 *)(this->songdata + pos));
+		}
+
+		/* After segments follows list of master tracks for each channel,
+		 * also prefixed with a byte counting actual tracks. */
+		loopmax = this->songdata[pos++];
+		for (loopidx = 0; loopidx < loopmax; loopidx++) {
+			/* Similar structure to segments list, but also has
+			 * the MIDI channel number as a byte before the offset
+			 * to next track. */
+			byte ch = this->songdata[pos++];
+			this->channels[ch].startpos = pos + 4;
+			pos += FROM_LE16(*(const int16 *)(this->songdata + pos));
+		}
+	}
+
+	/**
+	 * Read an SMF-style variable length value (note duration) from songdata.
+	 * @param pos Position to read from, updated to point to next byte after the value read
+	 * @return Value read from data stream
+	 */
+	uint16 ReadVariableLength(uint32 &pos)
+	{
+		byte b = 0;
+		uint16 res = 0;
+		do {
+			b = this->songdata[pos++];
+			res = (res << 7) + (b & 0x7F);
+		} while (b & 0x80);
+		return res;
+	}
+
+	/**
+	 * Prepare for playback from the beginning. Resets the song pointer for every track to the beginning.
+	 */
+	void RestartSong()
+	{
+		for (int ch = 0; ch < 16; ch++) {
+			Channel &chandata = this->channels[ch];
+			if (chandata.startpos != 0) {
+				/* Active track, set position to beginning */
+				chandata.playpos = chandata.startpos;
+				chandata.delay = this->ReadVariableLength(chandata.playpos);
+			} else {
+				/* Inactive track, mark as such */
+				chandata.playpos = 0;
+				chandata.delay = 0;
+			}
+		}
+	}
+
+	/**
+	 * Play one frame of data from one channel
+	 */
+	uint16 PlayChannelFrame(MidiFile::DataBlock &outblock, int channel)
+	{
+		uint16 newdelay = 0;
+		byte b1, b2;
+		Channel &chandata = this->channels[channel];
+
+		do {
+			/* Read command/status byte */
+			b1 = this->songdata[chandata.playpos++];
+
+			/* Command 0xFE, call segment from master track */
+			if (b1 == 0xFE) {
```

Comment:
Quite a lot of values in this function could do with naming, IMO - probably an enum

Reply:
On the other hand, those are values from a proprietary file format used in just a handful Microprose games of the era. They can't ever change and will only get used in this one place.

Reply:
I've changed the constants that are clearly from Standard MIDI into the existing enum values, but I don't think it's worth defining the "MPS MIDI" ones.